File: golang/hard-way/295.数据流的中位数.go
/*
 * @lc app=leetcode.cn id=295 lang=golang
 *
 * [295] 数据流的中位数
 *
 * https://leetcode-cn.com/problems/find-median-from-data-stream/description/
 *
 * algorithms
 * Hard (36.87%)
 * Likes:    77
 * Dislikes: 0
 * Total Accepted:    6.1K
 * Total Submissions: 15.4K
 * Testcase Example:  '["MedianFinder","addNum","addNum","findMedian","addNum","findMedian"]\n' +
  '[[],[1],[2],[],[3],[]]'
 *
 * 中位数是有序列表中间的数。如果列表长度是偶数，中位数则是中间两个数的平均值。
 *
 * 例如，
 *
 * [2,3,4] 的中位数是 3
 *
 * [2,3] 的中位数是 (2 + 3) / 2 = 2.5
 *
 * 设计一个支持以下两种操作的数据结构：
 *
 *
 * void addNum(int num) - 从数据流中添加一个整数到数据结构中。
 * double findMedian() - 返回目前所有元素的中位数。
 *
 *
 * 示例：
 *
 * addNum(1)
 * addNum(2)
 * findMedian() -> 1.5
 * addNum(3)
 * findMedian() -> 2
 *
 * 进阶:
 *
 *
 * 如果数据流中所有整数都在 0 到 100 范围内，你将如何优化你的算法？
 * 如果数据流中 99% 的整数都在 0 到 100 范围内，你将如何优化你的算法？
 *
 *
*/

// @lc code=start
type MedianFinder struct {
	Nums []int
}

func Constructor() MedianFinder {
	return MedianFinder{Nums: []int{}}

}

func (this *MedianFinder) AddNum(num int) {
	// 插入排序
	if len(this.Nums) == 0 || this.Nums[len(this.Nums)-1] <= num {
		this.Nums = append(this.Nums, num)
		return
	}

	var idx int
	for i := 0; i < len(this.Nums); i++ {
		if this.Nums[i] <= num && this.Nums[i+1] >= num {
			idx = i + 1
			break
		}
	}
	this.Nums = append(this.Nums, 0)
	copy(this.Nums[idx+1:], this.Nums[idx:])
	this.Nums[idx] = num
}

func (this *MedianFinder) FindMedian() float64 {
	n := len(this.Nums)
	mid := n / 2
	if n%2 == 1 {
		return float64(this.Nums[mid])
	}
	return float64(this.Nums[mid]+this.Nums[mid-1]) / 2.0
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * obj := Constructor();
 * obj.AddNum(num);
 * param_2 := obj.FindMedian();
 */
// @lc code=end

File: labuladong/数据结构/295.数据流的中位数.py
#
# @lc app=leetcode.cn id=295 lang=python3
#
# [295] 数据流的中位数
#

# @lc code=start
import heapq


class MinHeap:
    def __init__(self):
        self.nums = []

    @property
    def size(self):
        return len(self.nums)

    def pop(self):
        return heapq.heappop(self.nums)

    def peek(self):
        return heapq.nsmallest(1, self.nums)[0]

    def push(self, num):
        return heapq.heappush(self.nums, num)


class MaxHeap:
    def __init__(self):
        self.nums = []

    @property
    def size(self):
        return len(self.nums)

    def pop(self):
        return heapq.heappop(self.nums) * -1

    def peek(self):
        return heapq.nsmallest(1, self.nums)[0] * -1

    def push(self, num):
        return heapq.heappush(self.nums, num * -1)


class MedianFinder:
    def __init__(self):
        """
        将nums拆分为两个堆
            大顶堆：数据从小到大排列
            小顶堆: 数据从大到小排列

        每次加入数据时，选择两堆之一加入(优先往大顶堆里加)，
                并且保证
                1. 两个堆是有序的，并且两堆元素数量相差不超过1
                2. large堆的堆顶元素要大于等于small堆的堆顶元素
        这样中位数就是
            1.偶数 两堆堆顶的平均值
            2 奇数 元素数量多的那个堆顶元素
        """
        self.large = MinHeap()
        self.small = MaxHeap()

    def addNum(self, num: int) -> None:

        if self.small.size >= self.large.size:
            # small较大就要往large里加元素了
            self.small.push(num)
            self.large.push(self.small.pop())
        else:
            self.large.push(num)
            self.small.push(self.large.pop())

    def findMedian(self) -> float:
        if self.small.size == self.large.size:
            return (self.small.peek() + self.large.peek()) / 2.0
        elif self.small.size > self.large.size:
            return self.small.peek()
        else:
            return self.large.peek()

    @classmethod
    def test(cls):
        m = cls()
        m.addNum(1)
        m.addNum(2)
        m.addNum(3)
        print(m.findMedian())

        m = cls()
        m.addNum(-1)
        m.addNum(-2)
        m.addNum(-3)
        print(m.findMedian())


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()
# @lc code=end
